File: packages/adapters/nexus-adapter-sdks/nexus-adapter-sdk-go/define.go
package nexadapter

import (
	"context"
	"strings"
	"time"
)

type AdapterRuntimeContext struct {
	Context      context.Context
	Runtime      *RuntimeContext
	ConnectionID string
}

type AdapterContext[T any] struct {
	Context      context.Context
	Runtime      *RuntimeContext
	ConnectionID string
	Client       T
}

type ClientFactory[T any] struct {
	Create func(ctx AdapterRuntimeContext) (T, error)
}

type ConnectionHandlers[T any] struct {
	Connections func(ctx AdapterContext[T]) ([]AdapterConnectionIdentity, error)
	Health      func(ctx AdapterContext[T]) (*AdapterHealth, error)
}

type IngestHandlers[T any] struct {
	Monitor  func(ctx AdapterContext[T], emit EmitFunc) error
	Backfill func(ctx AdapterContext[T], since time.Time, emit EmitFunc) error
}

type SetupHandlers[T any] struct {
	Start  func(ctx AdapterContext[T], req AdapterSetupRequest) (*AdapterSetupResult, error)
	Submit func(ctx AdapterContext[T], req AdapterSetupRequest) (*AdapterSetupResult, error)
	Status func(ctx AdapterContext[T], req AdapterSetupRequest) (*AdapterSetupResult, error)
	Cancel func(ctx AdapterContext[T], req AdapterSetupRequest) (*AdapterSetupResult, error)
}

type DeclaredMethod[T any] struct {
	Description        string
	Action             string
	Params             map[string]any
	Response           map[string]any
	ConnectionRequired *bool
	MutatesRemote      *bool
	ContextHints       AdapterMethodContextHints
	Origin             *AdapterMethodOrigin
	Handler            func(ctx AdapterContext[T], req AdapterMethodRequest) (any, error)
}

type DefineAdapterConfig[T any] struct {
	Platform          string
	Name              string
	Version           string
	MultiAccount      bool
	CredentialService string
	Auth              *AdapterAuthManifest
	Capabilities      ChannelCapabilities
	MethodCatalog     *AdapterMethodCatalog
	Client            ClientFactory[T]
	Connection        ConnectionHandlers[T]
	Ingest            IngestHandlers[T]
	Setup             SetupHandlers[T]
	ServeStart        func(ctx context.Context, connectionID string, session *ServeSession) error
	Methods           map[string]DeclaredMethod[T]
}

func Method[T any](declaration DeclaredMethod[T]) DeclaredMethod[T] {
	return declaration
}

func DefineAdapter[T any](config DefineAdapterConfig[T]) Adapter {
	methods := config.Methods
	if methods == nil {
		methods = map[string]DeclaredMethod[T]{}
	}

	return Adapter{
		Operations: AdapterOperations{
			AdapterInfo: func(ctx context.Context) (*AdapterInfo, error) {
				info := buildAdapterInfo(config, methods)
				return &info, nil
			},
			AdapterConnectionsList: func(ctx context.Context) ([]AdapterConnectionIdentity, error) {
				adapterCtx, err := createAdapterContext(ctx, "", config.Client)
				if err != nil {
					return nil, err
				}
				if config.Connection.Connections != nil {
					return config.Connection.Connections(adapterCtx)
				}
				return defaultConnections(adapterCtx), nil
			},
			AdapterHealth: func(ctx context.Context, connectionID string) (*AdapterHealth, error) {
				adapterCtx, err := createAdapterContext(ctx, connectionID, config.Client)
				if err != nil {
					return nil, err
				}
				if config.Connection.Health != nil {
					health, err := config.Connection.Health(adapterCtx)
					if err != nil {
						return nil, err
					}
					if health != nil && strings.TrimSpace(health.ConnectionID) == "" {
						health.ConnectionID = adapterCtx.ConnectionID
					}
					return health, nil
				}
				return &AdapterHealth{
					Connected:    true,
					ConnectionID: adapterCtx.ConnectionID,
				}, nil
			},
			MonitorStart: func(ctx context.Context, connectionID string, emit EmitFunc) error {
				if config.Ingest.Monitor == nil {
					return nil
				}
				adapterCtx, err := createAdapterContext(ctx, connectionID, config.Client)
				if err != nil {
					return err
				}
				return config.Ingest.Monitor(adapterCtx, emit)
			},
			RecordsBackfill: func(ctx context.Context, connectionID string, since time.Time, emit EmitFunc) error {
				if config.Ingest.Backfill == nil {
					return nil
				}
				adapterCtx, err := createAdapterContext(ctx, connectionID, config.Client)
				if err != nil {
					return err
				}
				return config.Ingest.Backfill(adapterCtx, since, emit)
			},
			AdapterSetupStart: func(ctx context.Context, req AdapterSetupRequest) (*AdapterSetupResult, error) {
				if config.Setup.Start == nil {
					return nil, nil
				}
				adapterCtx, err := createAdapterContext(ctx, req.ConnectionID, config.Client)
				if err != nil {
					return nil, err
				}
				return config.Setup.Start(adapterCtx, req)
			},
			AdapterSetupSubmit: func(ctx context.Context, req AdapterSetupRequest) (*AdapterSetupResult, error) {
				if config.Setup.Submit == nil {
					return nil, nil
				}
				adapterCtx, err := createAdapterContext(ctx, req.ConnectionID, config.Client)
				if err != nil {
					return nil, err
				}
				return config.Setup.Submit(adapterCtx, req)
			},
			AdapterSetupStatus: func(ctx context.Context, req AdapterSetupRequest) (*AdapterSetupResult, error) {
				if config.Setup.Status == nil {
					return nil, nil
				}
				adapterCtx, err := createAdapterContext(ctx, req.ConnectionID, config.Client)
				if err != nil {
					return nil, err
				}
				return config.Setup.Status(adapterCtx, req)
			},
			AdapterSetupCancel: func(ctx context.Context, req AdapterSetupRequest) (*AdapterSetupResult, error) {
				if config.Setup.Cancel == nil {
					return nil, nil
				}
				adapterCtx, err := createAdapterContext(ctx, req.ConnectionID, config.Client)
				if err != nil {
					return nil, err
				}
				return config.Setup.Cancel(adapterCtx, req)
			},
			ServeStart:     config.ServeStart,
			Methods:        buildMethodHandlers(config, methods),
		},
	}
}

func createAdapterContext[T any](ctx context.Context, connectionID string, factory ClientFactory[T]) (AdapterContext[T], error) {
	runtimeCtx, err := loadRuntimeContextOptional()
	if err != nil {
		return AdapterContext[T]{}, err
	}

	resolvedConnectionID := strings.TrimSpace(connectionID)
	if resolvedConnectionID == "" && runtimeCtx != nil {
		resolvedConnectionID = strings.TrimSpace(runtimeCtx.ConnectionID)
	}

	adapterCtx := AdapterContext[T]{
		Context:      ctx,
		Runtime:      runtimeCtx,
		ConnectionID: resolvedConnectionID,
	}

	if factory.Create != nil {
		client, err := factory.Create(AdapterRuntimeContext{
			Context:      ctx,
			Runtime:      runtimeCtx,
			ConnectionID: resolvedConnectionID,
		})
		if err != nil {
			return AdapterContext[T]{}, err
		}
		adapterCtx.Client = client
	}

	return adapterCtx, nil
}

func loadRuntimeContextOptional() (*RuntimeContext, error) {
	runtimeCtx, err := LoadRuntimeContextFromEnv()
	if err == nil {
		return runtimeCtx, nil
	}
	if strings.Contains(err.Error(), "missing runtime context") {
		return nil, nil
	}
	return nil, err
}

func defaultConnections[T any](ctx AdapterContext[T]) []AdapterConnectionIdentity {
	connectionID := strings.TrimSpace(ctx.ConnectionID)
	if connectionID == "" && ctx.Runtime != nil {
		connectionID = strings.TrimSpace(ctx.Runtime.ConnectionID)
	}
	if connectionID == "" {
		return []AdapterConnectionIdentity{}
	}

	connection := AdapterConnectionIdentity{
		ID:     connectionID,
		Status: "ready",
	}
	if ctx.Runtime != nil && ctx.Runtime.Credential != nil {
		if ref := strings.TrimSpace(ctx.Runtime.Credential.Ref); ref != "" {
			connection.CredentialRef = ref
		}
	}
	return []AdapterConnectionIdentity{connection}
}

func buildMethodHandlers[T any](config DefineAdapterConfig[T], methods map[string]DeclaredMethod[T]) map[string]func(ctx context.Context, req AdapterMethodRequest) (any, error) {
	if len(methods) == 0 {
		return nil
	}
	built := make(map[string]func(ctx context.Context, req AdapterMethodRequest) (any, error), len(methods))
	for name, declaration := range methods {
		decl := declaration
		built[name] = func(ctx context.Context, req AdapterMethodRequest) (any, error) {
			adapterCtx, err := createAdapterContext(ctx, req.ConnectionID, config.Client)
			if err != nil {
				return nil, err
			}
			return decl.Handler(adapterCtx, req)
		}
	}
	return built
}

func buildAdapterInfo[T any](config DefineAdapterConfig[T], methods map[string]DeclaredMethod[T]) AdapterInfo {
	operations := []AdapterOperation{
		OpAdapterInfo,
		OpAdapterConnectionsList,
		OpAdapterHealth,
	}
	if config.Ingest.Monitor != nil {
		operations = append(operations, OpAdapterMonitorStart)
	}
	if config.Ingest.Backfill != nil {
		operations = append(operations, OpRecordsBackfill)
	}
	if config.Setup.Start != nil {
		operations = append(operations, OpAdapterSetupStart)
	}
	if config.Setup.Submit != nil {
		operations = append(operations, OpAdapterSetupSubmit)
	}
	if config.Setup.Status != nil {
		operations = append(operations, OpAdapterSetupStatus)
	}
	if config.Setup.Cancel != nil {
		operations = append(operations, OpAdapterSetupCancel)
	}
	if config.ServeStart != nil {
		operations = append(operations, OpAdapterServeStart)
	}

	methodDescriptors := make([]AdapterMethod, 0, len(methods))
	for name, declaration := range methods {
		methodDescriptors = append(methodDescriptors, buildMethodDescriptor(config, name, declaration))
	}

	info := AdapterInfo{
		Platform:             config.Platform,
		Name:                 config.Name,
		Version:              config.Version,
		Operations:           operations,
		Methods:              methodDescriptors,
		CredentialService:    config.CredentialService,
		MultiAccount:         config.MultiAccount,
		PlatformCapabilities: config.Capabilities,
		Auth:                 config.Auth,
	}
	if len(methodDescriptors) > 0 {
		if config.MethodCatalog != nil {
			info.MethodCatalog = config.MethodCatalog
		} else {
			info.MethodCatalog = &AdapterMethodCatalog{
				Source:    "manifest",
				Namespace: config.Platform,
			}
		}
	} else if config.MethodCatalog != nil {
		info.MethodCatalog = config.MethodCatalog
	}
	return info
}

func buildMethodDescriptor[T any](config DefineAdapterConfig[T], name string, declaration DeclaredMethod[T]) AdapterMethod {
	action := strings.TrimSpace(declaration.Action)
	if action == "" {
		if declaration.MutatesRemote != nil && *declaration.MutatesRemote {
			action = "write"
		} else {
			action = "read"
		}
	}
	connectionRequired := true
	if declaration.ConnectionRequired != nil {
		connectionRequired = *declaration.ConnectionRequired
	}
	mutatesRemote := action == "write"
	if declaration.MutatesRemote != nil {
		mutatesRemote = *declaration.MutatesRemote
	} else if action == "write" {
		mutatesRemote = true
	}
	origin := declaration.Origin
	if origin == nil {
		origin = &AdapterMethodOrigin{
			PackageID:         config.Platform,
			PackageVersion:    config.Version,
			DeclarationMode:   "builtin",
			DeclarationSource: "sdk.define",
			Namespace:         config.Platform,
		}
	}

	contextHints := declaration.ContextHints
	if contextHints.Params == nil {
		contextHints.Params = map[string]AdapterMethodContextHintValue{}
	}

	return AdapterMethod{
		Name:               name,
		Description:        declaration.Description,
		Action:             action,
		Params:             declaration.Params,
		Response:           declaration.Response,
		ConnectionRequired: connectionRequired,
		MutatesRemote:      mutatesRemote,
		Origin:             *origin,
		ContextHints:       contextHints,
	}
}

File: packages/adapters/nexus-adapter-sdks/nexus-adapter-sdk-ts/src/define.ts
import type {
  AdapterConnectionIdentity,
  AdapterAuthManifest,
  AdapterHealth,
  AdapterInboundRecord,
  AdapterInfo,
  AdapterMethod,
  AdapterMethodCatalog,
  AdapterMethodContextHints,
  AdapterOperation,
  AdapterSetupResult,
  ChannelCapabilities,
} from "./protocol.js";
import type {
  AdapterContext,
  AdapterDefinition,
  AdapterMethodInvokeRequest,
  AdapterOperations,
  AdapterSetupRequest,
} from "./run.js";

type MaybePromise<T> = T | Promise<T>;

export type DefinedAdapterContext<TClient> = AdapterContext & {
  connectionId?: string;
  client: TClient | undefined;
};

export type DeclaredAdapterMethod<TClient> = {
  description?: string | null;
  action?: "read" | "write";
  params?: Record<string, unknown> | null;
  response?: Record<string, unknown> | null;
  connection_required?: boolean;
  mutates_remote?: boolean;
  context_hints?: AdapterMethodContextHints;
  handler: (
    ctx: DefinedAdapterContext<TClient>,
    req: AdapterMethodInvokeRequest,
  ) => MaybePromise<unknown>;
};

export type DefineAdapterConfig<TClient> = {
  platform: string;
  name: string;
  version: string;
  multi_account?: boolean;
  credential_service?: string;
  auth?: AdapterAuthManifest;
  capabilities: ChannelCapabilities;
  methodCatalog?: AdapterMethodCatalog;
  client?: {
    create?: (args: { ctx: AdapterContext; connectionId?: string }) => MaybePromise<TClient>;
  };
  connection?: {
    connections?: (ctx: DefinedAdapterContext<TClient>) => MaybePromise<AdapterConnectionIdentity[]>;
    health?: (ctx: DefinedAdapterContext<TClient>) => MaybePromise<Omit<AdapterHealth, "connection_id">>;
  };
  ingest?: {
    monitor?: (
      ctx: DefinedAdapterContext<TClient>,
      emit: (record: AdapterInboundRecord) => void,
    ) => MaybePromise<void>;
    backfill?: (
      ctx: DefinedAdapterContext<TClient>,
      args: { since: Date },
      emit: (record: AdapterInboundRecord) => void,
    ) => MaybePromise<void>;
  };
  setup?: {
    start?: (
      ctx: DefinedAdapterContext<TClient>,
      req: AdapterSetupRequest,
    ) => MaybePromise<AdapterSetupResult>;
    submit?: (
      ctx: DefinedAdapterContext<TClient>,
      req: AdapterSetupRequest,
    ) => MaybePromise<AdapterSetupResult>;
    status?: (
      ctx: DefinedAdapterContext<TClient>,
      req: AdapterSetupRequest,
    ) => MaybePromise<AdapterSetupResult>;
    cancel?: (
      ctx: DefinedAdapterContext<TClient>,
      req: AdapterSetupRequest,
    ) => MaybePromise<AdapterSetupResult>;
  };
  serve?: AdapterOperations["adapter.serve.start"];
  methods?: Record<string, DeclaredAdapterMethod<TClient>>;
};

export function method<TClient>(
  declaration: DeclaredAdapterMethod<TClient>,
): DeclaredAdapterMethod<TClient> {
  return declaration;
}

export function defineAdapter<TClient = unknown>(
  config: DefineAdapterConfig<TClient>,
): AdapterDefinition {
  const methodDefinitions = config.methods ?? {};

  const operations: AdapterOperations = {
    "adapter.info": async () => buildAdapterInfo(config, methodDefinitions),
    "adapter.connections.list": async (ctx) => {
      const definedCtx = await createDefinedContext(config, ctx, ctx.runtime?.connection_id);
      if (config.connection?.connections) {
        return await config.connection.connections(definedCtx);
      }
      return defaultConnections(ctx);
    },
    "adapter.health": async (ctx, args) => {
      const definedCtx = await createDefinedContext(config, ctx, args.connection_id);
      if (config.connection?.health) {
        const result = await config.connection.health(definedCtx);
        return {
          ...result,
          connection_id: args.connection_id,
        };
      }
      return {
        connected: true,
        connection_id: args.connection_id,
      };
    },
    methods: buildMethodHandlers(config, methodDefinitions),
  };

  if (config.ingest?.monitor) {
    operations["adapter.monitor.start"] = async (ctx, args, emit) => {
      const definedCtx = await createDefinedContext(config, ctx, args.connection_id);
      await config.ingest!.monitor!(definedCtx, emit);
    };
  }

  if (config.ingest?.backfill) {
    operations["records.backfill"] = async (ctx, args, emit) => {
      const definedCtx = await createDefinedContext(config, ctx, args.connection_id);
      await config.ingest!.backfill!(definedCtx, { since: args.since }, emit);
    };
  }

  if (config.setup?.start) {
    operations["adapter.setup.start"] = async (ctx, req) => {
      const definedCtx = await createDefinedContext(config, ctx, req.connection_id);
      return await config.setup!.start!(definedCtx, req);
    };
  }
  if (config.setup?.submit) {
    operations["adapter.setup.submit"] = async (ctx, req) => {
      const definedCtx = await createDefinedContext(config, ctx, req.connection_id);
      return await config.setup!.submit!(definedCtx, req);
    };
  }
  if (config.setup?.status) {
    operations["adapter.setup.status"] = async (ctx, req) => {
      const definedCtx = await createDefinedContext(config, ctx, req.connection_id);
      return await config.setup!.status!(definedCtx, req);
    };
  }
  if (config.setup?.cancel) {
    operations["adapter.setup.cancel"] = async (ctx, req) => {
      const definedCtx = await createDefinedContext(config, ctx, req.connection_id);
      return await config.setup!.cancel!(definedCtx, req);
    };
  }

  if (config.serve) {
    operations["adapter.serve.start"] = config.serve;
  }

  return { operations };
}

async function createDefinedContext<TClient>(
  config: DefineAdapterConfig<TClient>,
  ctx: AdapterContext,
  connectionId?: string,
): Promise<DefinedAdapterContext<TClient>> {
  const client = config.client?.create ? await config.client.create({ ctx, connectionId }) : undefined;
  return {
    ...ctx,
    connectionId,
    client,
  };
}

function defaultConnections(ctx: AdapterContext): AdapterConnectionIdentity[] {
  const connectionId = ctx.runtime?.connection_id?.trim();
  if (!connectionId) {
    return [];
  }
  return [
    {
      id: connectionId,
      status: "ready",
      ...(ctx.runtime?.credential?.ref ? { credential_ref: ctx.runtime.credential.ref } : {}),
    },
  ];
}

function buildMethodHandlers<TClient>(
  config: DefineAdapterConfig<TClient>,
  methods: Record<string, DeclaredAdapterMethod<TClient>>,
): AdapterOperations["methods"] {
  if (Object.keys(methods).length === 0) {
    return {};
  }
  return Object.fromEntries(
    Object.entries(methods).map(([name, declaration]) => [
      name,
      async (ctx: AdapterContext, req: AdapterMethodInvokeRequest) => {
        const definedCtx = await createDefinedContext(config, ctx, req.connection_id);
        return await declaration.handler(definedCtx, req);
      },
    ]),
  );
}

function buildAdapterInfo<TClient>(
  config: DefineAdapterConfig<TClient>,
  methods: Record<string, DeclaredAdapterMethod<TClient>>,
): AdapterInfo {
  const operations: AdapterOperation[] = [
    "adapter.info",
    "adapter.connections.list",
    "adapter.health",
    ...(config.ingest?.monitor ? (["adapter.monitor.start"] as const) : []),
    ...(config.ingest?.backfill ? (["records.backfill"] as const) : []),
    ...(config.setup?.start ? (["adapter.setup.start"] as const) : []),
    ...(config.setup?.submit ? (["adapter.setup.submit"] as const) : []),
    ...(config.setup?.status ? (["adapter.setup.status"] as const) : []),
    ...(config.setup?.cancel ? (["adapter.setup.cancel"] as const) : []),
    ...(config.serve ? (["adapter.serve.start"] as const) : []),
  ];

  const methodDescriptors = Object.entries(methods).map(([name, declaration]) =>
    buildMethodDescriptor(config, name, declaration),
  );

  return {
    platform: config.platform,
    name: config.name,
    version: config.version,
    operations,
    methods: methodDescriptors,
    ...(config.credential_service ? { credential_service: config.credential_service } : {}),
    multi_account: config.multi_account ?? false,
    platform_capabilities: config.capabilities,
    ...(config.auth ? { auth: config.auth } : {}),
    ...(methodDescriptors.length > 0
      ? {
          methodCatalog: config.methodCatalog ?? {
            source: "manifest",
            namespace: config.platform,
          },
        }
      : config.methodCatalog
        ? { methodCatalog: config.methodCatalog }
        : {}),
  };
}

function buildMethodDescriptor<TClient>(
  config: DefineAdapterConfig<TClient>,
  name: string,
  declaration: DeclaredAdapterMethod<TClient>,
): AdapterMethod {
  return {
    name,
    ...(declaration.description !== undefined ? { description: declaration.description } : {}),
    ...(declaration.action ? { action: declaration.action } : {}),
    ...(declaration.params ? { params: declaration.params } : {}),
    ...(declaration.response ? { response: declaration.response } : {}),
    connection_required: declaration.connection_required ?? true,
    mutates_remote:
      typeof declaration.mutates_remote === "boolean"
        ? declaration.mutates_remote
        : declaration.action === "write",
    context_hints: declaration.context_hints ?? { params: {} },
    origin: {
      package_kind: "adapter",
      package_id: config.platform,
      package_version: config.version,
      declaration_mode: "manifest",
      declaration_source: "package declaration",
      namespace: config.platform,
    },
  } as AdapterMethod;
}
